code/forms/LDAPChangePasswordForm.php

```diff
@@ -0,0 +1,127 @@
+<?php
+
+class LDAPChangePasswordForm extends ChangePasswordForm {
+
+	/**
+	 * Change the password
+	 *
+	 * @param array $data The user submitted data
+	 * @return SS_HTTPResponse
+	 */
+	public function doChangePassword(array $data) {
+
+		/**
+		 * @var LDAPService $service
+		 */
+		$service = Injector::inst()->get('LDAPService');
+
+		if($member = Member::currentUser()) {
+			try {
+				$userData = $service->getUserByGUID($member->GUID);
+			} catch(Exception $e) {
+				SS_Log::log($e->getMessage(), SS_Log::ERR);
+				$this->clearMessage();
+				$this->sessionMessage(
+					_t('LDAPAuthenticator.PROBLEMFINDINGDATA', 'There was a problem retrieving your user data'),
+					'bad'
+				);
+				return $this->controller->redirect($this->controller->Link('changepassword'));
+			}
+			$loginResult = $service->authenticate($userData['cn'], $data['OldPassword']);
+			if(!$loginResult['success']) {
+				$this->clearMessage();
+				$this->sessionMessage(
+					_t('Member.ERRORPASSWORDNOTMATCH', "Your current password does not match, please try again"),
+					"bad"
+				);
+				// redirect back to the form, instead of using redirectBack() which could send the user elsewhere.
+				return $this->controller->redirect($this->controller->Link('changepassword'));
+			}
+		}
+
+		if(!$member) {
+			if(Session::get('AutoLoginHash')) {
+				$member = Member::member_from_autologinhash(Session::get('AutoLoginHash'));
+			}
+
+			// The user is not logged in and no valid auto login hash is available
+			if(!$member) {
+				Session::clear('AutoLoginHash');
+				return $this->controller->redirect($this->controller->Link('login'));
+			}
+		}
+
+		// Check the new password
+		if(empty($data['NewPassword1'])) {
+			$this->clearMessage();
+			$this->sessionMessage(
+				_t('Member.EMPTYNEWPASSWORD', "The new password can't be empty, please try again"),
+				"bad");
+
+			// redirect back to the form, instead of using redirectBack() which could send the user elsewhere.
+			return $this->controller->redirect($this->controller->Link('changepassword'));
+		}
+		else if($data['NewPassword1'] == $data['NewPassword2']) {
+			try {
+				$isValid = $service->setPassword($member, $data['NewPassword1']);
+			} catch(Exception $e) {
+				SS_Log::log($e->getMessage(), SS_Log::ERR);
+				$this->clearMessage();
+				$this->sessionMessage(
+					_t('LDAPAuthenticator.PROBLEMFINDINGDATA', 'There was a problem retrieving your user data'),
```

Comment:
I'd probably try to tell the user what they should do: is it a transient failure and they should retry, or it's most likely fatal and they should get in touch with the administrator?